main/src/ca/uwaterloo/flix/language/phase/EntryPoint.scala

```diff
@@ -0,0 +1,295 @@
+/*
+ * Copyright 2022 Matthew Lutze
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.{Ast, Scheme, SourceLocation, Symbol, Type, TypedAst}
+import ca.uwaterloo.flix.language.errors.EntryPointError
+import ca.uwaterloo.flix.language.phase.unification.ClassEnvironment
+import ca.uwaterloo.flix.util.Validation.{ToFailure, ToSuccess, flatMapN, mapN}
+import ca.uwaterloo.flix.util.{InternalCompilerException, Validation}
+
+/**
+  * Processes the entry point of the program.
+  *
+  * The entry point is replaced by a new function (`main%`) that calls the old entry point (`func`).
+  *
+  * If the argument to `func` has type `Array[String]`,
+  *   then the arguments are passed from `main%` to `func`.
+  * If the argument to `func` has type `Unit`,
+  *   then the arguments to `main%` are ignored.
+  * If the argument to `func` has some other type,
+  *   an error is raised.
+  *
+  * If the result type of `func` has type `Unit`,
+  *   then the result is returned from `main%` as normal.
+  * If the result type of `func` is some other type with a `ToString` instance,
+  *   then the result is printed in `main%`.
+  * If the result type of `func` is some other type without a `ToString` instance,
+  *   then an error is raised.
+  *
+  *  For example, given an entry point `func` with type `Array[String] -> Float64`,
+  *  we produce:
+  *  {{{
+  *  pub def main%(args: Array[String]): Unit = {
+  *      println(func(args))
+  *  }
+  *  }}}
+  */
+object EntryPoint {
+
+  /**
+    * The resulting scheme of the entry point function.
+    * `Array[String] -> Unit`
+    */
+  private val EntryPointScheme = Scheme(Nil, Nil, Type.mkImpureArrow(Type.mkArray(Type.Str, SourceLocation.Unknown), Type.Unit, SourceLocation.Unknown))
+
+  /**
+    * The default entry point in case none is specified. (`main`)
+    */
+  private val DefaultEntryPoint = Symbol.mkDefnSym("main")
+
+
+  def run(root: TypedAst.Root)(implicit flix: Flix): Validation[TypedAst.Root, EntryPointError] = flix.phase("EntryPoint") {
+    flatMapN(findOriginalEntryPoint(root)) {
+      // Case 1: We have an entry point. Wrap it.
```

Comment:
Comment goes after Some=>

Usually we have the None case before the Some case (Maybe add to STYLE.md)